Revisions #2: Revisions state update and diff-ing

bperson opened the pull request:
This is the second chunk coming from my WIP PR to introduce post revisions in calypso: #13367. It pushes the normalisation of revisions a bit further and adds support for diff-ing between revisions

### changes
It builds upon what was introduced in #12733, adding:

- Normalise dates from a format with timezone information in the field names (`date_gmt`) to a format with timezone information embedded in the datetime strings (simply adding the zone designator for the zero UTC offset: `Z`). This means we can keep strings in Redux ( instead of moment.js objects) and still [use components](https://github.com/Automattic/wp-calypso/blob/master/client/reader/post-time/index.jsx#L37) parsing our strings without any timezone issue.
- Add display/editing normalisation updates
- Introduce [jsdiff](https://github.com/kpdecker/jsdiff) as a dependency (no shrinkmap yet, #14157 need to be merged first).

We compute the diff between revisions:
- when loading the revisions from the API, saving the changes summary ( words added / words removed ) directly in Redux
- when displaying a revision in the diff viewer (coming in the last PR of this serie), it re-computes the complete diff to display all changes, memoized via the `getPostRevisionChanges` selector.

### testing

Besides unit tests, this is currently used nowhere. You can see how it can be used in the WIP PR #13367, for instance in the [diff viewer](https://github.com/Automattic/wp-calypso/blob/724642f09cbbec14699fbe753960eef17198d24a/client/post-editor/editor-diff-viewer/index.jsx#L47).

bperson commented:
This has been rebased on the latest `add/fetch-users`, I still need to get #13828 merged first to provide `requestUsers` but most of it is standing up on its own.

cc @aduth @nylen or if you know anyone else that might be interested in reviewing (mostly) tests ;)

youknowriad commented on client/state/data-layer/wpcom/posts/revisions/index.js:
```diff
@@ -64,8 +74,17 @@ export const receiveError = ( { dispatch }, { siteId, postId }, next, rawError )
+	forEach( normalizedRevisions, ( revision, index ) => {
+		revision.changes = index === normalizedRevisions.length - 1
+			? { added: 0, removed: 0 }
+			: countDiffWords( diffWords( normalizedRevisions[ index + 1 ].content, revision.content ) );
```
Curious what's the reasoning behind computing the diff when fetching the revisions? Naively, I'd have let this for a memoized selector, for example, assuming this can be heavy and compute the diff only when needed.

Maybe not needed but what if we want to diff an old revision with the last one for example.

bperson replied:
I went this way precisely because it's quite expansive [0]: persisting the summary results along the revisions speeds up the first draw of the revisions list after a reload compared to the equivalent memoized-selector-based solution. That said, it's also partially un-normalizing the revision data in Redux as you get redundant data (the summary can be recomputed from the contents).

> Maybe not needed but what if we want to diff an old revision with the last one for example.

The ability to diff between any 2 revisions is dropped for now: https://calypsop2.wordpress.com/2017/04/19/editor-revisions-i3/#comment-7039
That said, for those features outside the "main" one, it's quite easy to recompute the diff between two revisions with the utils functions introduced here.

As an interesting footnote here: we can most likely go even further and only update the Redux state if we receive deleted or new revisions if we make the assumption that revisions are immutable, AFAICS, there is no way of editing a revision from the API or in wp-admin.

[0] Myers algorithm (used by jsdiff) has a pretty interesting time complexity for revisions like object: it's in O(ND) with N the sum of the number of words in revisions A and B, and D the size of the shortest edit sequence A -> B. If we suppose that most revisions are incremental updates on a post, D should stay small :)

youknowriad commented on client/state/data-layer/wpcom/posts/revisions/index.js:
```diff
@@ -64,8 +74,17 @@ export const receiveError = ( { dispatch }, { siteId, postId }, next, rawError )
 	dispatch( receivePostRevisionsSuccess( siteId, postId ) );
-	dispatch( receivePostRevisions( siteId, postId, map( revisions, normalizeRevision ) ) );
+	dispatch( receivePostRevisions( siteId, postId, normalizedRevisions ) );
+	dispatch( requestUsers( siteId, uniq( map( normalizedRevisions, 'author' ) ) ) );
```
Do we really want to query the users as soon as we receive the revisions? My first thought is that we should probably use a query component for this when needed. Is there a particular reason for this to be fetched here?

bperson replied:
The initial reason for it to be in there was to bundle revisions and users by dispatching `receivePostRevisions`, `receivePostRevisionsSuccess` and `receiveUsers` in the same tick. But I dropped it, from a UX pov, having the author names is nice but not as important as accessing the revisions asap: it's better to have a first render of the revisions while we request more information about the various authors imho.

I've updated this PR to remove this and will probably bundle `QueryUsers` with its "usage" in the last PR for revisions.

youknowriad commented on package.json:
```diff
@@ -48,6 +48,7 @@
     "creditcards": "2.1.2",
     "cross-env": "4.0.0",
     "debug": "2.2.0",
+    "diff": "1.4.0",
```
Should we update the lock https://wpcalypso.wordpress.com/devdocs/docs/shrinkwrap.md?term=shrinkw ?

youknowriad replied:
Nevermind, I just saw your note above

bperson replied:
Since this is not depending on #13367 anymore, this can be merged on its own, I've updated the shrinkmap :)

bperson commented:
### Updated
- removed all usage of `requestUsers` in favor of using a `QueryUsers` component alongside `QueryPostRevisions` in the last PR for revisions
- it's rebased on the latest `master`. It doesn't directly need anything introduced by `add/fetch-users` anymore, so I updated the shrinkmap, making this mergeable on its own, no issue running with the last shrinkmap locally.

aduth commented on client/lib/text-utils/index.js:
```diff
@@ -1,3 +1,9 @@
+/**
+ * External Dependencies
+ */
+import JsDiff from 'diff';
```
Can we `import { diffWords } from 'diff';` here instead of accessing property below?

bperson replied:
we definitely can :) fixed

aduth commented on client/lib/text-utils/index.js:
```diff
@@ -24,4 +30,26 @@ function countWords( content ) {
 }
 
-export default { countWords };
+const diffWords = JsDiff.diffWords;
```
Minor: Destructuring could help avoid some repetition: `const { diffWords } = JsDiff;`

bperson replied:
removed it completely as we're directly importing (and re-exporting in the "default" object `diffWords`)

aduth commented on client/lib/text-utils/index.js:
```diff
@@ -24,4 +30,26 @@ function countWords( content ) {
+		diffChanges,
+		( accumulator, change ) => {
+			const count = countWords( change.value );
+			change.added && ( accumulator.added += count );
```
Minor: I found myself staring at this function for minutes convinced we'd be inaccurately incrementing _both_ `added` and `removed` for each change, until I noticed that the first half of each line is treated effectively as a condition. This is one case where I'd advocate for the less magical `if ( change.added ) {` variation.

bperson replied:
I've no strong feelings about either so I updated it to use `ifs`

aduth commented on client/lib/text-utils/index.js:
```diff
@@ -24,4 +30,26 @@ function countWords( content ) {
+	);
+}
+
+export default {
```
Minor and not related to these changes specifically: Exporting only a default of an object makes it difficult to [tree shake](https://webpack.js.org/guides/tree-shaking/) (dead code elimination) unused functions within a file. We're not yet on Webpack 2 (#10799, #11204) so this isn't entirely relevant yet anyways, but in anticipation it's typically better to export/import functions individually.

bperson replied:
TIL :) 

I don't really want to touch the imports of `countWords` if at all possible in this PR ;) Keeping it on my todo once this PR is merged though.

bperson replied:
Created a new PR to make sure we don't drop this: #14794

it's based on this one so we should merge this one first.

aduth commented on client/state/data-layer/wpcom/posts/revisions/index.js:
```diff
@@ -31,10 +33,15 @@ export function normalizeRevision( revision ) {
+		...flow(
+			pick( [ 'date_gmt', 'modified_gmt' ] ),
+			mapValues( val => `${ val }Z` ),
+			mapKeys( key => key.slice( 0, -'_gmt'.length ) )
```
Minor: Readability: I think it's nice that this isn't a magic number, but at the same time the `-` is a bit foreign at first glance. Wondering if either `-1 * '_gmt'.length` or `-4 /* "_gmt" */` would be more readable. What do you think?

bperson replied:
if we want to truly explain why the -1: `-4 /* truncate "_gmt" */`? I don't see any way to give the "truncate" meaning to the `-1` in code?

aduth commented on client/state/posts/revisions/selectors.js:
```diff
@@ -1,21 +1,74 @@
 /**
```
All new selectors should ideally be added to the `client/state/selectors` directory. `selectors.js` is a legacy pattern.

https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#selectors

bperson replied:
fixed

aduth commented on client/state/posts/revisions/selectors.js:
```diff
@@ -1,21 +1,74 @@
+
+	return {
+		...revision,
+		...{ author }
```
This can be simplified as just `author`.

```js
return {
	...revision,
	author
};
```

bperson replied:
fixed

aduth commented on client/state/data-layer/wpcom/posts/revisions/index.js:
```diff
@@ -31,10 +33,15 @@ export function normalizeRevision( revision ) {
 			mapValues( ( val = {} ) => val.rendered )
+		)( revision ),
+		...flow(
+			pick( [ 'date_gmt', 'modified_gmt' ] ),
```
I'm bewildered by how this is working, specifically the flow/currying applied to `pick`. In my own Node console it errors as I'd expect it to, because `pick` isn't supposed to return a function but rather the object subset:

```
n_ > _.flow( _.pick( 'foo' ) )( { foo: 'bar' } )
TypeError: Expected a function
```

Can you explain what I'm missing?

bperson replied:
first guess would be: `_.pick` is the non-fp version? which means it returns an empty object `{}`, and `flow` doesn't like non-functions as an arg?

bperson replied:
well, I'm not answering the question in fact :)

Lodash's fp follows a few "defaults", one of them is to re-order arguments in a "data-last" fashion: `_.pick( { foo: 'bar' }, 'foo' )` === fpPick( 'foo' )( { foo: 'bar' } )`

aduth replied:
But `pick` here is imported from `lodash`, not `lodash/fp`.

And to this point, I don't think we'd want to reintroduce `lodash/fp`, as we'd made a point previously to try to remove it (#11354).

aduth replied:
Oh, it is imported from `lodash/fp`. I totally missed that. Hmm!

bperson replied:
Thinking a bit more about it, I haven't found any usage of Lodash's fp in calypso (no import, no usage inside a flow AFAICS), it might not be a good idea to start importing it (even more so if people aren't familiar with it since it can cause confusions like this one ;) and the [FP docs aren't amazing](https://github.com/lodash/lodash/wiki/FP-Guide) :( )

bperson replied:
well didn't refresh :) Will fix it to use the "classical" one

aduth replied:
Yeah, I actually quite like how this looks here. It's nice to compose the operations this way. But I'd agree we might not want to set this precedent for risk of confusing developers with the presence of both variants (i.e. being consistent on using one vs. the other).
